### ISSUE TITLE
SINDy for finding algebraic relations rather than ODEs

### ISSUE THREAD

**Issue by LasseLedet** [dynamicslab/pysindy#228]
Hi,

Can I trick SINDy to find me sparse relations of algebraic equations? 

For example, I have a set of measured data x, y, z and I want to find the sparsest relation that relates x to y and z i.e. find x = f(y,z), where f is likely on polynomial form in both y and z. 

Note that x, y and z are not functions of time, but just measured values. Hope it makes sense :-)

**Comment by akaptano** [comment]
Yes! We play tricks like this with the code all the time. Simply pass a pre-defined vector "x_dot_train" to model.fit(x_train, t=t_train, x_dot=x_dot_train), which will tell the code to avoid computing x_dot and just use whatever vector you have passed to the code. So instead of passing the derivative of [x, y, z], just pass the data itself.

**Comment by LasseLedet** [comment]
Hi Alan,

Thanks for your reply – and glad to hear that SINDy can help me.

Playing around with it I still don’t seem to  get it, though. 

I have a function H(Q) – just as a polynomial for starters – where I would like just to recover the polynomial coefficients. 

 

I am doing the following



 

Then



 

Where do I go wrong?

 

Kind regards/Med venlig hilsen,

 

Lasse Ledet

**Comment by LasseLedet** [comment]
Hi Alan,

Sorry for the late reply - The holiday caught up with me :-)

Let me try again :-) 

I have a function H(Q) – just as a polynomial for starters – where I would like just to recover the polynomial coefficients. 

I am doing the following
`Q = np.linspace(0,10,1000)
H = 0.44*Q**2 + 0.25*Q + 20
x_dot_train = np.ones(len(H))
x_train2 = np.stack((H,Q),axis=1)`

Then
`poly_order = 5
threshold = 0.05

model = ps.SINDy(
    optimizer=ps.STLSQ(threshold=threshold),
    feature_library=ps.PolynomialLibrary(degree=poly_order))
model.fit(x_train2, x_dot = x_dot_train)
model.print()`

Where do I go wrong?

**Comment by akaptano** [comment]
It is still not totally clear to me what the goal is here. You seem to be doing [1, 1] = Theta(up to 5th order polynomials in H and Q) * coefficients, so you will generate fits like H^5 + 3 * Q = 1 which maybe is the type of polynomial you want?

But from your description it sounded like you wanted to discover the coefficients of H. In this case, you should just pass H as "x_dot_train" and Q as "x_train2", since what you are trying to do is H = Theta(polynomials of Q) * coefficients. So model.fit should now return H(Q) = 0.44 * Q ** 2 + 0.25 * Q + 20. Of course, you only need a second order polynomial feature library for this.

**Comment by LasseLedet** [comment]
Thanks a lot - It works now 🙌🙌🙌
